File: solutions/javascript/resistor-color-trio/1/resistor-color-trio.js
export class ResistorColorTrio {
  color_list = [
  'black',
  'brown',
  'red',
  'orange',
  'yellow',
  'green',
  'blue',
  'violet',
  'grey',
  'white'
  ]
  
  prefixes = {
    "giga": 10**9,
    "mega": 10**6,
    "kilo": 10**3
  }
  
  constructor(colors) {
    this.colors = colors;
  }

  #decode(color) {
    const decoded = this.color_list.indexOf(color);
    if (decoded === -1) throw new Error("invalid color");
    
    return this.color_list.indexOf(color);
  }

  get label() {
    const firstDigit = this.#decode(this.colors[0]);
    const secondDigit = this.#decode(this.colors[1]);
    const multiplier = this.#decode(this.colors[2]);
  
    let num = (firstDigit*10 + secondDigit) * 10 ** multiplier;
    let prefix = "";
        
    for (const key in this.prefixes) {  
        const magnitude = this.prefixes[key];
        if (num > magnitude && num % magnitude == 0) {
            num /= magnitude;
            prefix = key;
            break;
        }
    }
    
    return `Resistor value: ${num} ${prefix}ohms`;
  }
}

File: solutions/python/resistor-color-trio/1/resistor_color_trio.py
COLORS = {
    "black": 0,
    "brown": 1,
    "red": 2,
    "orange": 3,
    "yellow": 4,
    "green": 5,
    "blue": 6,
    "violet": 7,
    "grey": 8,
    "white": 9
}

PREFIXES = {
    10**9: "giga",
    10**6: "mega",
    10**3: "kilo"
}

def label(colors):
    num = (10 * COLORS[colors[0]] + COLORS[colors[1]]) * 10 ** COLORS[colors[2]]
    prefix = ""
    
    for magnitude in PREFIXES:
        if num > magnitude and num % magnitude == 0:
            num //= magnitude
            prefix = PREFIXES[magnitude]
            break
    
    return f"{num} {prefix}ohms"
